### PR TITLE
Make '-webkit-background-size' a legacy shorthand

### PR DESCRIPTION
**Pull request by pxlcoder**
#### 856140e9a06f431fae7c3df24443aa963ea8445a
<pre>
Make &apos;-webkit-background-size&apos; a legacy shorthand
<a href="https://bugs.webkit.org/show_bug.cgi?id=243562">https://bugs.webkit.org/show_bug.cgi?id=243562</a>
rdar://98151497

Reviewed by Tim Nguyen.

`background-size` and `-webkit-background-size` are a pair of properties that
share a computed style. `-webkit-background-size` differs from `background-size`
in the interpretation of a single value: `-webkit-background-size: l;` is
equivalent to `background-size: l l;`, whereas `background-size: l;` is
equivalent to `background-size: l auto;`.

Property pairs that share a computed style but differ in syntax should be
implemented as legacy shorthands.

Additionally, always serialize `background-size` and `-webkit-background-size`
as two values, so that values can be roundtripped between the prefixed and
unprefixed variant.

Spec: <a href="https://www.w3.org/TR/css-cascade-5/#aliasing">https://www.w3.org/TR/css-cascade-5/#aliasing</a>

* LayoutTests/TestExpectations:

Remove expectations for a forms test that was unexpectedly setting
`-webkit-background-size`, as it was not a shorthand.

* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style.html:
* LayoutTests/fast/backgrounds/size/parsing-background-size-values-expected.txt:
* LayoutTests/fast/backgrounds/size/parsing-inherit-expected.txt:
* LayoutTests/fast/backgrounds/size/resources/parsing-background-size-values.js:
* LayoutTests/fast/backgrounds/size/resources/parsing-inherit.js:
* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/KeyframeEffect-getKeyframes.tentative.html:

Update test to use two-value syntax.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/background-332-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/background-size-001.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-size-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/quirks/unitless-length/excluded-properties-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/quirks/unitless-length/excluded-properties-002.html:

Include an additional check for `1234px auto`, to avoid a false positive, as `background-size` is serialized to two values.

* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/svg/css/getComputedStyle-basic-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fillSizeToCSSValue):

Return two values from `getComputedStyle` unless the value is &quot;cover&quot;, &quot;contain&quot;, or &quot;auto&quot;.

(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeBackgroundSize):

Always serialize `background-size` and `-webkit-background-size` as two values.

(WebCore::CSSPropertyParser::parseSingleValue):
(WebCore::CSSPropertyParser::parseShorthand):

Canonical link: <a href="https://commits.webkit.org/254925@main">https://commits.webkit.org/254925@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/53250114608a2ac6a4fb0e1a58df1555e03e54f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21180 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99874 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158232 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33625 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28812 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82906 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96309 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26761 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77387 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26608 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81539 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81346 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69628 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34724 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15383 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32539 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16361 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3440 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36304 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39280 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38218 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35450 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/css/StyleProperties.cpp [hunk 1]
```diff
@@ -376,6 +376,8 @@ String StyleProperties::getPropertyValue(CSSPropertyID propertyID, Document* doc
         return get2Values(scrollPaddingInlineShorthand());
     case CSSPropertyWebkitTextOrientation:
         return getPropertyValue(CSSPropertyTextOrientation);
+    case CSSPropertyWebkitBackgroundSize:
+        return getPropertyValue(CSSPropertyBackgroundSize);
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp:380]: That's not good, there is no round-trip:

```js
element.style.backgroundSize = "100px"; // behaves as "100px auto"
element.style.webkitBackgroundSize; // "100px", oops!
element.style.webkitBackgroundSize += ""; // changes behavior to "100px 100px"
```

```js
element.style.webkitBackgroundSize = "100px"; // behaves as "100px 100px"
element.style.backgroundSize; // "100px", oops!
element.style.backgroundSize += ""; // changes behavior to "100px auto"
```

And I guess this was already wrong, but please fix gCS while you are at it:

```js
element.style.backgroundSize = "100px";
getComputedStyle(element).webkitBackgroundSize; // "100px", oops!
```

This is OK:
```js
element.style.webkitBackgroundSize = "100px";
getComputedStyle(element).backgroundSize; // "100px 100px"
```

See https://github.com/whatwg/compat/issues/28, either serialize `background-size` and `-webkit-background-size` differently, or always use 2 values. People there seemed to prefer 2 values.

File: Source/WebCore/css/CSSProperties.json [hunk 2]
```diff
@@ -5371,7 +5370,9 @@
         },
         "-webkit-background-size": {
             "codegen-properties": {
-                "related-property": "background-size",
+                "longhands": [
+                    "background-size"
+                ],
                 "name-for-methods": "Size",
                 "fill-layer-property": true
```
Comment (Loirooriol) [inline Source/WebCore/css/CSSProperties.json]: I think `name-for-methods` and `fill-layer-property` are not relevant for shorthands.

**Review by Loirooriol — CHANGES_REQUESTED** [review]

File: Source/WebCore/css/parser/CSSPropertyParser.cpp [hunk 3]
```diff
@@ -6560,6 +6559,13 @@ bool CSSPropertyParser::parseShorthand(CSSPropertyID property, bool important)
     }
     case CSSPropertyBackground:
         return consumeBackgroundShorthand(backgroundShorthand(), important);
+    case CSSPropertyWebkitBackgroundSize: {
+        RefPtr<CSSValue> backgroundSize = consumeCommaSeparatedBackgroundComponent(CSSPropertyWebkitBackgroundSize, m_range, m_context.mode);
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParser.cpp]: nit: `RefPtr backgroundSize` or `auto`

File: Source/WebCore/css/StyleProperties.cpp (hunk 1 above)
Comment (zcorpan) [reply Source/WebCore/css/StyleProperties.cpp:380]: Yes, and make the unprefixed property parse like -webkit-background-size if it's web-compatible. https://github.com/whatwg/compat/issues/28#issuecomment-700427017

Comment (pxlcoder) [reply Source/WebCore/css/StyleProperties.cpp:380]: Changing the parsing is out-of-scope for this change, and would carry more web compat risk.

I have changed serialization to always use two values.

File: Source/WebCore/css/parser/CSSPropertyParser.cpp [hunk 4]
```diff
@@ -3239,25 +3239,39 @@ static RefPtr<CSSPrimitiveValue> consumeBackgroundSize(CSSPropertyID property, C
...
-        if (range.peek().id() == CSSValueAuto) // `auto' is the default
-            range.consumeIncludingWhitespace();
-        else
+        vertical = consumeIdent<CSSValueAuto>(range);
+        if (!vertical)
             vertical = consumeLengthOrPercent(range, cssParserMode, ValueRange::NonNegative, UnitlessQuirk::Allow);
-    } else if (!vertical && property == CSSPropertyWebkitBackgroundSize) {
-        // Legacy syntax: "-webkit-background-size: 10px" is equivalent to "background-size: 10px 10px".
-        vertical = horizontal;
+        if (range.atEnd() && !vertical)
```
Comment (Loirooriol) [inline Source/WebCore/css/parser/CSSPropertyParser.cpp]: Can this happen? If `!vertical` it means that both `consumeIdent` and `consumeLengthOrPercent` failed, so nothing should have been consumed and `!range.atEnd()` should still hold?

Comment (pxlcoder) [reply Source/WebCore/css/parser/CSSPropertyParser.cpp]: Good catch, I have removed this logic.

**Review by Loirooriol — APPROVED** [review]
Informal LGTM.
It would be good to fix getComputedStyle while you are at it, but it's already broken so not blocking on that.

**Comment by nt1m** [comment]
> It would be good to fix getComputedStyle while you are at it, but it's already broken so not blocking on that.

I think Aditya's patch makes sure to always serialize to 2 values, so that would also get fixed, I would think.

We could add a test for it.

**Comment by Loirooriol** [comment]
No, `getComputedStyle` still omits the 2nd value if it's `auto`, which is wrong for `-webkit-background-size`. See https://searchfox.org/wubkat/rev/7b1a4e7e5fb469b0eea41fe131f2f190d08f2383/Source/WebCore/css/ComputedStyleExtractor.cpp#2115-2116
```cpp
    if (fillSize.size.height.isAuto())
        return ComputedStyleExtractor::zoomAdjustedPixelValueForLength(fillSize.size.width, style);
```

**Comment by nt1m** [comment]
> No, `getComputedStyle` still omits the 2nd value if it's `auto`, which is wrong for `-webkit-background-size`. See https://searchfox.org/wubkat/rev/7b1a4e7e5fb469b0eea41fe131f2f190d08f2383/Source/WebCore/css/ComputedStyleExtractor.cpp#2115-2116
> 
> ```c++
>     if (fillSize.size.height.isAuto())
>         return ComputedStyleExtractor::zoomAdjustedPixelValueForLength(fillSize.size.width, style);
> ```

Good catch! Seems easily fixable, we could pass in a `CSSPropertyID` for `fillSizeToCSSValue` and only do this adjustment for the gCS for mask-size

**Review by nt1m — APPROVED** [review]
r=me with tests on EWS changed to match the new expected results (change the expected results to keep them passing)

Also, please do a WPT export.

**Comment by pxlcoder** [comment]
Following the change to gCS, the following WPTs are failing:

```
imported/w3c/web-platform-tests/css/css-backgrounds/background-332.html
imported/w3c/web-platform-tests/css/css-backgrounds/background-size-001.html
imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-size-computed.html
imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo.html
```

They expect a single value for `background-size`, when a single value is specified. If I were to update the tests to match our new behavior, they would fail on other browsers. Using `assert_any` would fix that issue, but also seems a little strange.

@nt1m @zcorpan Any suggestions on how to proceed with gCS / these WPT?

**Comment by nt1m** [comment]
@pxlcoder You can either change the expected result, or use `assert_in_array()` to accept both forms. I slightly favor the first option to make other browsers follow suit (or participate in the discussion for https://github.com/whatwg/compat/issues/28), and getting interoperable behavior.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 5]
```diff
@@ -2110,15 +2110,15 @@ static Ref<CSSValue> maskModeToCSSValue(MaskMode type)
     return CSSValuePool::singleton().createValue(CSSValueMatchSource);
 }
 
-static Ref<CSSValue> fillSizeToCSSValue(const FillSize& fillSize, const RenderStyle& style)
+static Ref<CSSValue> fillSizeToCSSValue(const FillSize& fillSize, const RenderStyle& style, CSSPropertyID propertyID)
```
Comment (nt1m) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: Can you make the propertyID the first argument to be consistent with ` createPositionListForLayer` and `createSingleAxisPositionValueForLayer`?

**Review by nt1m — APPROVED** [review]
Can you add a GTK specific baseline for:
LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo.html ?

This is actually not a regression of your patch but of https://github.com/WebKit/WebKit/commit/a3ee3aaf6b9225948517d86782fed2ea39802032 , but would be nice to fix anyway here since your patch touches the baseline.

**Comment by nt1m** [comment]
iOS needs a similar rebaseline for LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo.html

**Comment by pxlcoder** [comment]
> iOS needs a similar rebaseline for LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo.html

Not sure what's going on here – the patch contains a baseline in `LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt`, and the test is passing locally 🤔

**Comment by pxlcoder** [comment]
(turns out there's already a baseline in `ios-wk2`, will update that one instead of adding a new one)

**Pull request merged**